source/sqlsrv/stmt.cpp

```diff
@@ -1772,40 +1818,25 @@ void fetch_fields_common( _Inout_ ss_sqlsrv_stmt* stmt, _In_ zend_long fetch_typ
 	}
 
-	// get the numer of columns in the result set
-	SQLSMALLINT num_cols = core::SQLNumResultCols(stmt TSRMLS_CC);
+    // get the numer of columns in the result set
+    SQLSMALLINT num_cols = get_resultset_meta_data(stmt);
 
 	// if this is the first fetch in a new result set, then get the field names and
 	// store them off for successive fetches.
-	if(( fetch_type & SQLSRV_FETCH_ASSOC ) && stmt->fetch_field_names == NULL ) {
+    if ((fetch_type & SQLSRV_FETCH_ASSOC) && stmt->fetch_field_names == NULL) {
 
         SQLLEN field_name_len = 0;
-        SQLSMALLINT field_name_len_w = 0;
-        SQLWCHAR field_name_w[( SS_MAXCOLNAMELEN + 1 ) * 2] = {L'\0'};
-        sqlsrv_malloc_auto_ptr<char> field_name;
         sqlsrv_malloc_auto_ptr<sqlsrv_fetch_field_name> field_names;
-        field_names = static_cast<sqlsrv_fetch_field_name*>( sqlsrv_malloc( num_cols * sizeof( sqlsrv_fetch_field_name )));
-        SQLSRV_ENCODING encoding = (( stmt->encoding() == SQLSRV_ENCODING_DEFAULT ) ? stmt->conn->encoding() : stmt->encoding());
-        for( int i = 0; i < num_cols; ++i ) {
-
-            core::SQLColAttributeW ( stmt, i + 1, SQL_DESC_NAME, field_name_w, ( SS_MAXCOLNAMELEN + 1 ) * 2, &field_name_len_w, NULL TSRMLS_CC );
-
-            //Conversion function expects size in characters
-            field_name_len_w = field_name_len_w / sizeof ( SQLWCHAR );
-            bool converted = convert_string_from_utf16( encoding, field_name_w,
-                field_name_len_w, ( char** ) &field_name, field_name_len );
-
-            CHECK_CUSTOM_ERROR( !converted, stmt, SQLSRV_ERROR_FIELD_ENCODING_TRANSLATE, get_last_error_message() ) {
-                throw core::CoreException();
-            }
-
-            field_names[i].name = static_cast<char*>( sqlsrv_malloc( field_name_len, sizeof( char ), 1 ));
-            memcpy_s(( void* )field_names[i].name, ( field_name_len * sizeof( char )) , ( void* ) field_name, field_name_len );
-            field_names[i].name[field_name_len] = '\0';  // null terminate the field name since SQLColAttribute doesn't.
-            field_names[i].len = field_name_len + 1;
-            field_name.reset();
+        field_names = static_cast<sqlsrv_fetch_field_name*>(sqlsrv_malloc(num_cols * sizeof(sqlsrv_fetch_field_name)));
+        for (int i = 0; i < num_cols; ++i) {
+            // The meta data field name is already null-terminated, and the field name len is correct.
+            field_name_len = stmt->current_meta_data[i]->field_name_len;
+            field_names[i].name = static_cast<char*>(sqlsrv_malloc(field_name_len, sizeof(char), 1));
+            memcpy_s((void*)field_names[i].name, (field_name_len * sizeof(char)), (void*)stmt->current_meta_data[i]->field_name, field_name_len);
+            field_names[i].name[field_name_len] = '\0';     // null terminate the field name after the memcpy
+            field_names[i].len = field_name_len;            // field_name_len should not need to include the null char
```

Comment:
The old code seems to correct for wide char encoding. Is that no longer necessary with your changes?

Reply:
Because when getting the meta data the field names have been converted if necessary

Reply:
In core_sqlsrv_field_metadata?